src/cryptenroll/cryptenroll-pkcs11.c

```diff
@@ -36,27 +34,9 @@ int enroll_pkcs11(
         if (r < 0)
                 return r;
 
-        pkey = X509_get0_pubkey(cert);
-        if (!pkey)
-                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to extract public key from X.509 certificate.");
-
-        r = rsa_pkey_to_suitable_key_size(pkey, &decrypted_key_size);
-        if (r < 0)
-                return log_error_errno(r, "Failed to determine RSA public key size.");
-
-        log_debug("Generating %zu bytes random key.", decrypted_key_size);
-
-        decrypted_key = malloc(decrypted_key_size);
-        if (!decrypted_key)
-                return log_oom();
-
-        r = crypto_random_bytes(decrypted_key, decrypted_key_size);
-        if (r < 0)
-                return log_error_errno(r, "Failed to generate random key: %m");
-
-        r = rsa_encrypt_bytes(pkey, decrypted_key, decrypted_key_size, &encrypted_key, &encrypted_key_size);
+        r = X509_certificate_generate_volume_key(cert, &decrypted_key, &decrypted_key_size, &savedata, &savedata_size);
         if (r < 0)
-                return log_error_errno(r, "Failed to encrypt key: %m");
+                return log_error_errno(r, "Failed to generate secret key for X.509 certifcate: %m");
```

Comment:
certifcate is a typo.

so here you do duplicate logging. your new helper x509_certificate_generate_volume_key() already logs about all errors, hence you shouldn't do so here again.

In systemd we have this rule that all functions should by one of two kinds: either they log about all errors on their own, or they log about no errors, and propagate it instead, so that the caller can log. For each function determine which typo it should be, and then stick to this rule. Typically, "main program"-like functions are of the logging type and "library"-like functions about the non-logging type. (logging at LOG_DEBUG level is always OK, it does not count as logging in this context).

By following this rule we ensure that we neither log twice about errors nor not at all.

Here, X509_certificate_generate_volume_key() is of the logging kind, hence you should drop the exta logging here, and just do `return r`.